stats/statsview.cpp

```diff
@@ -80,17 +80,40 @@ void StatsView::mouseReleaseEvent(QMouseEvent *)
 	}
 }
 
+class RootNode : public QSGNode
+{
+public:
+	RootNode();
+	QSGImageNode *imageNode; // imageNode to plot QGRaphicsScene on. Remove in due course.
+	// We entertain one node per Z-level.
+	std::array<QSGNode *, (size_t)ChartZValue::Count> zNodes;
+	std::array<std::vector<ChartItem *>, (size_t)ChartZValue::Count> items;
+};
+
+RootNode::RootNode()
+{
+	for (QSGNode *&zNode: zNodes) {
+		zNode = new QSGNode;
+		appendChildNode(zNode);
+	}
+}
+
 QSGNode *StatsView::updatePaintNode(QSGNode *oldNode, QQuickItem::UpdatePaintNodeData *)
 {
 	// The QtQuick drawing interface is utterly bizzare with a distinct 1980ies-style memory management.
 	// This is just a copy of what is found in Qt's documentation.
-	QSGImageNode *n = static_cast<QSGImageNode *>(oldNode);
-	if (!n)
-		n = rootNode = window()->createImageNode();
+	RootNode *n = static_cast<RootNode *>(oldNode);
+	if (!n) {
+		n = rootNode = new RootNode;
+		n->imageNode = window()->createImageNode();
+		n->zNodes[(int)ChartZValue::Series]->appendChildNode(n->imageNode);
```

Comment:
so the nature of 'z' is simply a side effect of the order of these nodes, not something that the Qt code actually understands as a concept - or am I misunderstanding how this is supposed to work?

Reply:
This is right. We have n top-level nodes, which are rendered according to creation time and to these we add the child nodes at the various z-levels. The effect is that nodes with a lower z-value are rendered below nodes with a higher z-value by the shader. Inside a z-level the nodes are rendered according to creation time. Therefore, this simulates setZValue of QGraphicsScene.

Reply:
Yes, that's what I figured. So instead of an 'actual' Z value implementation in the Qt infrastructure we are simulating Z value using that 'fast infrastructure' in our implementation. Which is fine, and I like the naming that you put in place there.